File: python/Greedy/maximum_subarray.py
from typing import List

class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        """
        Find the contiguous subarray with the largest sum (Kadane's Algorithm).
        
        Approach:
        - Initialize maxSub to the first element (handles all-negative arrays).
        - Use curSum to track the current subarray sum.
        - For each number:
          - If curSum is negative, reset it to 0 (start a new subarray).
          - Add the current number to curSum.
          - Update maxSub if curSum is greater than maxSub.
        - Return maxSub as the result.
        
        Time Complexity: O(N), where N is the length of nums
        - We iterate through the array once
        
        Space Complexity: O(1)
        - Only a constant amount of extra space is used
        
        Args:
            nums: List[int] - List of integers
        Returns:
            int - Maximum subarray sum
        """
        maxSub = nums[0]
        curSum = 0

        for n in nums:
            if curSum < 0:
                curSum = 0  # Start a new subarray if the current sum is negative
            curSum += n
            maxSub = max(maxSub, curSum)
        return maxSub

# -------------------
# Test Cases
# -------------------
def test_max_subarray():
    s = Solution()
    # Test case 1: Mixed positive and negative
    assert s.maxSubArray([-2,1,-3,4,-1,2,1,-5,4]) == 6, "Failed on mixed array"
    # Test case 2: All positive
    assert s.maxSubArray([1,2,3,4,5]) == 15, "Failed on all positive"
    # Test case 3: All negative
    assert s.maxSubArray([-1,-2,-3,-4]) == -1, "Failed on all negative"
    # Test case 4: Single element
    assert s.maxSubArray([5]) == 5, "Failed on single element"
    # Test case 5: Single negative element
    assert s.maxSubArray([-5]) == -5, "Failed on single negative element"
    # Test case 6: Large negative followed by positives
    assert s.maxSubArray([-10,2,3,4]) == 9, "Failed on large negative followed by positives"
    # Test case 7: Zeros and positives
    assert s.maxSubArray([0,0,0,1]) == 1, "Failed on zeros and positives"
    # Test case 8: Zeros and negatives
    assert s.maxSubArray([0,0,0,-1]) == 0, "Failed on zeros and negatives"
    print("All test cases passed!")

if __name__ == "__main__":
    test_max_subarray()

File: C++/Greedy/maximum_subarray.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    /**
     * Function to find the contiguous subarray with the largest sum.
     *
     * Approach: Kadane's Algorithm
     * - Initialize maxSub to the first element (handles arrays with all negative numbers)
     * - Use curSum to keep track of the current subarray sum.
     * - If curSum becomes negative, reset it to 0 (start a new subarray)
     * - Iterate through the array, updating maxSub whenever curSum is larger.
     *
     * Time Complexity: O(N) — iterate through the array once
     * Space Complexity: O(1) — constant extra space
     */
    int maxSubArray(vector<int>& nums) {
        int maxSub = nums[0];  // Stores maximum subarray sum found so far
        int curSum = 0;        // Current subarray sum

        for (int n : nums) {
            if (curSum < 0)
                curSum = 0;   // Start a new subarray if current sum < 0
            curSum += n;      // Add current number to the running sum
            maxSub = max(maxSub, curSum);  // Update maxSub if needed
        }

        return maxSub;
    }
};

/*
====================================================
Line-by-line C++ syntax explanation:
====================================================

#include <bits/stdc++.h>
    -> Includes all standard C++ libraries (like <vector>, <iostream>, etc.)

using namespace std;
    -> Allows using standard library names (like vector, string) without prefixing them with std::

class Solution {
public:
    -> Defines a class named Solution with public member functions.

int maxSubArray(vector<int>& nums) {
    -> Defines a member function `maxSubArray` that takes a reference to a vector of integers and returns an int.

int maxSub = nums[0];
    -> Initialize `maxSub` to the first element of the array. This is the best sum found so far.

int curSum = 0;
    -> Initialize `curSum` to 0. This will store the sum of the current subarray.

for (int n : nums) {
    -> Loop over each element `n` in the vector `nums`.

if (curSum < 0)
    curSum = 0;
    -> If the current subarray sum becomes negative, reset it to 0 to start a new subarray.

curSum += n;
    -> Add the current number to the current subarray sum.

maxSub = max(maxSub, curSum);
    -> Update the maximum sum found so far if the current sum is higher.

return maxSub;
    -> Return the maximum subarray sum found.

====================================================
Time Complexity:
    - O(N): We iterate over the array exactly once.

Space Complexity:
    - O(1): We only use two integer variables regardless of input size.
====================================================
*/

